### ISSUE TITLE
Audit: 54 AiModelBuilder Configure* methods silently drop their configuration

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1876]
## Summary

**54 of 114 `Configure*` methods on `AiModelBuilder` are fully inert**: every private field they assign is never read anywhere in the codebase. The fluent API accepts the configuration, returns `this`, and silently drops it. Callers get no exception, no warning — just a model trained as if they had never called the method.

This is not only an observability gap. Several are **correctness** bugs: you can configure a loss function or a learning-rate scheduler and training will ignore it.

## Verified examples

**`ConfigureLossFunction` — silently ignored.** The training loop reads the model's default instead:

```csharp
// AiModelBuilder.BuildPipeline.cs:314
var lossFunction = _model.DefaultLossFunction;
```
```csharp
// AiModelBuilder.CoreML.cs:44-48  — assigned, never read
public IAiModelBuilder<T, TInput, TOutput> ConfigureLossFunction(ILossFunction<T> lossFunction)
{
    _configuredLossFunction = lossFunction;
    return this;
}
```
`grep -rn "_configuredLossFunction" src/` returns exactly two lines: the declaration and that assignment. Configure `CrossEntropy` on a regression-default model and it trains with MSE.

**`ConfigureLearningRateScheduler` — no-op.** `_configuredLearningRateScheduler` (`AiModelBuilder.CoreML.cs:29, 207`) is assigned and never read; `ILearningRateScheduler` appears in no training path.

**`ConfigureStoppingCriterion` — no-op.** `_configuredStoppingCriterion` (`AiModelBuilder.Coverage.cs:39, 301`) assigned, never read. Combined with the absence of default early stopping, there is currently **no way to stop a fit early through the facade** for most model families.

**`ConfigureCrossValidation` / `ConfigureDataSplitter` / `ConfigureActivationFunction`** — same shape.

## Full list — fully inert (every assigned field never read)

<details><summary>54 methods</summary>

ConfigureActivationFunction, ConfigureActiveLearning, ConfigureAdversarialAttack, ConfigureAdversarialDefense, ConfigureAnomalyDetector, ConfigureAudioEffect, ConfigureAudioEnhancer, ConfigureBenchmark, ConfigureCertifiedDefense, ConfigureClassificationMetric, ConfigureClusterMetric, ConfigureClustering, ConfigureContinualLearning, ConfigureCrossValidation, ConfigureCurriculumScheduler, ConfigureDataSplitter, ConfigureDataTransformer, ConfigureDistanceMetric, ConfigureDistillationStrategy, ConfigureDocumentModel, ConfigureDocumentStore, ConfigureDriftDetection, ConfigureEmbeddingModel, ConfigureEnvironment, ConfigureExplorationStrategy, ConfigureExternalClusterMetric, ConfigureFinancialModel, ConfigureGaussianProcess, ConfigureInterpolation, ConfigureInterpolation2D, ConfigureKernelFunction, ConfigureLayer, ConfigureLearningRateScheduler, ConfigureLinkFunction, ConfigureLossFunction, ConfigureMatrixDecomposition, ConfigureModelCompressionStrategy, ConfigureModelExplainer, ConfigureModelOptions, ConfigureNoiseScheduler, ConfigurePDESpecification, ConfigurePointCloudModel, ConfigureQueryStrategy, ConfigureRadialBasisFunction, ConfigureRegressionMetric, ConfigureSSLMethod, ConfigureScoringRule, ConfigureSimilarityMetric, ConfigureStoppingCriterion, ConfigureTimeSeriesDecomposition, ConfigureTool, ConfigureVideoModel, ConfigureWaveletFunction, ConfigureWindowFunction

</details>

**Partially wired** (some assigned fields dead, method still does something):

| method | dead fields | live fields |
|---|---|---|
| `ConfigureFederatedLearning` | 8 | 6 |
| `ConfigureHyperparameterOptimizer` | 1 (`_hyperparameterTrials`) | 2 |
| `ConfigurePipelineParallelism` | 1 (`_pipelineMicroBatchCount`) | 3 |

**Fully wired:** 57 methods. So roughly **47% of the configuration surface is inert.**

## Related: the same bug class in training callbacks

#1875 fixes an instance of this for the TimeSeries family: `ConfigureTrainingCallback` documents *"return `false` to request an early stop"*, but those models run their own epoch loop inside `Train()` (which the facade calls once), so the callback fired **once, after training had already finished** — `facade observed 1 epoch(s); expected one per training epoch`. Nothing could be observed and no veto could take effect.

## Method

```
1. parse every `public IAiModelBuilder<...> Configure*(...)` body in src/AiModelBuilder*.cs
2. collect the private fields it assigns
3. count reads across all of src/ (excluding src/Generated) as:
       occurrences - assignments - declaration
4. a method is "fully inert" when every field it assigns has 0 reads
```

**Caveats — this is triage, not proof.** Reads via reflection or source generators would not be detected (`src/Generated` was excluded; `YamlConfigApplier.g.cs` does reference `ILearningRateScheduler`, though it can only call the same inert setter). An earlier, sloppier pass of this audit reported 66 and produced false positives — `ConfigureFitDetector` and `ConfigureFitnessCalculator` **are** read and were wrongly flagged. Each entry should be confirmed individually before it is fixed. The five named in "Verified examples" above were each checked by hand.

## Proposed triage

1. **Training-critical first** (correctness): `ConfigureLossFunction`, `ConfigureLearningRateScheduler`, `ConfigureStoppingCriterion`, `ConfigureCrossValidation`, `ConfigureDataSplitter`, `ConfigureActivationFunction`, plus the metric methods.
2. **Industry-standard defaults on**: early stopping and LR scheduling should be enabled by default for any model, not merely configurable. For reference, the ecosystem's defaults are weaker — sklearn's `MLPClassifier` defaults `early_stopping=False`, and Keras's `EarlyStopping` defaults `restore_best_weights=False` — so on-by-default with best-weight restore is a deliberate step past them.
3. **Everything else**: decide per method — wire it, or make it throw `NotSupportedException` so it fails loudly instead of lying. A method that cannot be honored should not silently accept input.
4. **Guard the class of bug**: a test (or analyzer) asserting every `Configure*` field has a reader would stop this regressing.

Related: #1875.

**Comment by ooples** [comment]
## Follow-up: `ConfigureLossFunction` is dead at **three** layers, not one

Worth recording before anyone attempts the "just wire the field up" fix, because the one-line version is a trap.

**Layer 1 — the facade field.** `_configuredLossFunction` is assigned and never read (`AiModelBuilder.CoreML.cs:19, 46`).

**Layer 2 — the facade only uses a loss for *reporting*.** The `lossFunction` resolved at `BuildPipeline.cs:314` has exactly one consumer:

```csharp
// AiModelBuilder.BuildPipeline.cs:579  — the ONLY use
var loss = lossFunction.CalculateLoss(predictionVector, targetVector);
```

That feeds the epoch-loss metric. The **gradient** does not come from it — `model.Train(batch)` dispatches through `TrainWithTape` → the model's *own* loss:

```csharp
// NeuralNetworks/NeuralNetworkBase.cs:927, :990
var lossResult = LossFunction.CalculateLossAndGradientGpu(output, target);
```

So changing line 314 to `_configuredLossFunction ?? _model.DefaultLossFunction` would make the **reported** loss honor the caller while training still optimized the model's built-in loss — arguably worse than today's honest-but-ignored behavior, since the number in the log would no longer describe what was optimized. (I wrote that one-liner, then reverted it for exactly this reason.)

**Layer 3 — no seam exists to set a model's loss.** `NeuralNetworkBase.LossFunction` is `protected` and assigned only in the constructor (`:210, :432, :447`). There is no `SetLossFunction`, no `ISupportsLossFunction`. And on the TimeSeries side, `TimeSeriesRegressionOptions.LossFunction` (`:127`) is a settable option that **no TimeSeries model ever reads** — another dead knob one layer down.

### Implication

Honoring `ConfigureLossFunction` requires a real seam, roughly:

1. an interface (e.g. `ISupportsLossFunction<T>`) letting the facade inject a loss before training;
2. `NeuralNetworkBase` implementing it (assign `LossFunction`, guarding against mid-training mutation);
3. `TimeSeriesModelBase` implementing it, **and** the TimeSeries models actually consuming `Options.LossFunction` instead of their hardcoded losses;
4. the facade's reporting loss and the model's training loss resolving to the same instance, so the logged number always describes what was optimized.

This pattern likely repeats for other tier-1 entries (`ConfigureActivationFunction` almost certainly has the same constructor-only problem). **The audit's read-count is a lower bound on the work: a field having a reader does not mean the value reaches training.**

**Comment by ooples** [comment]
## Full tiered audit committed: `docs/FACADE_CONFIG_AUDIT.md`

On branch `fix/timeseries-early-stopping` (#1875) — the complete surface, tiered, with `file:line` for every entry: Tier 1 training-critical, **Tier 1b unwireable-by-construction**, Tier 2 metrics, Tier 3 workflows, Tier 4 domain models, Tier 5 math primitives, Tier 6 robustness, plus the 3 PARTIAL methods.

Two findings since the original post that change the shape of the work:

### The purged splitters already exist AND already derive `DataSplitterBase<T>`

Every splitter in `Preprocessing/DataPreparation/Splitting/TimeSeries/` — `PurgedKFoldSplitter`, `CombinatorialPurgedSplitter`, `AnchoredWalkForwardSplitter`, `WalkForwardSplitter`, `RollingOriginSplitter`, `TimeSeriesSplitter`, `SlidingWindowSplitter`, `BlockedTimeSeriesSplitter` — already derives `DataSplitterBase<T>`, i.e. already satisfies `IDataSplitter<T>`, the exact type `_configuredDataSplitter` holds.

**Wiring that one dead field unlocks ~58 splitters at once, purged ones included.** That makes `ConfigureDataSplitter` the highest-leverage item in the audit: the capability that would put this library ahead of every mainstream framework is already written, tested, and simply unreachable.

### New Tier 1b: implementations that can't be wired because they skip the base class

A `Configure*` method takes an interface. An implementation that is a `static class` can never be passed to it, no matter what the builder does. Fixing the dead field is necessary but not sufficient.

| gap | consequence |
|---|---|
| `Finance/Evaluation/PurgedWalkForwardValidator.cs:27` is a **`static class`** | It is the **only** implementation carrying `labelHorizon` + `embargo`, and it can never reach `ConfigureDataSplitter`. Should derive `DataSplitterBase<T>` (or be wrapped by a subclass that does). |
| **No `ICrossValidator` implements purge/embargo** | `TimeSeriesCrossValidator.cs:29` is rolling-origin only. Once `ConfigureCrossValidation` is wired there will still be **nothing purged to pass it**. Needs a `CrossValidatorBase` subclass wrapping the purged fold logic. |
| `ConfigureLossFunction` takes `ILossFunction<T>`, but tape training requires `LossFunctionBase<T>` (`NeuralNetworkBase.cs:6887`) | All 40 in-repo losses derive the base, so in-repo use is safe — but a user-supplied interface implementation throws deep inside training. Narrow the parameter or validate at configure time. |

**Rule of thumb going in the doc:** if a capability is worth configuring, its implementations must derive from the contract's base class. A parallel static API guarantees the facade's version drifts from the real one — which is precisely how the purged splitters ended up correct and unusable.

**Comment by ooples** [comment]
Triage note (PR #1789 sweep): reviewed against the #1789 branch. There is related activity on that branch but **no verified fix for what this issue actually asks for**, so it is deliberately left open and not linked for auto-close. Re-check after #1789 merges.
